### PR TITLE
fix(dialog): allow click events to propagate

### LINKED ISSUE #794: mdc-dialog with list items ; how to listen for clicks properly?
I'd like to use the dialog with list items. Of course the purpose is for a user to select an item. See example on official site "choose a ringtone": https://material-components-web.appspot.com/dialog.html

However, I noticed that it's not possible by default for my list items to listen for clicks. It seems that the framework attaches an event handler on the dialog and stops propagation, not letting me any chance to handle the selection of an item.

```
$(document).ready(function() {
  const MDCDialog = mdc.dialog.MDCDialog;
  const dialog = new MDCDialog(document.getElementById('mdc-dialog-with-list'));
  dialog.show();
  
  $(document).on('click', '.mdc-list-item', function() {
    alert('hello');
  });
});
```
https://jsfiddle.net/ftmn974x/

I guess I need to write a new adapter for the foundation to let a click happen in a dialog? But that seems way too heavy.

What I don't get as well is that in the official example and guideline this kind of list is clearly a list of items that you can select. However,I have to implement the css when hovering an item myself as well as the event handlers (and as mentioned above maybe even rewrite an adapter for this simple issue?).

Thank you for any advice

### PR DESCRIPTION
**Pull request by jonathanzong**
This PR fixes #794. I implemented the short fix proposed by @amsheehan in that thread and confirmed that listeners on the list items receive click events. cc @jverkoey

### CODE REVIEW

**Comment by amsheehan** [comment]
Hey @jonathanzong, Thanks for the PR!

It looks like some of the tests are failing. As soon as that gets fixed up I'll review. From the output in Travis CI, it looks like there are two tests failing:

`foundation.test.js:199`
&
`foundation.test.js:256`

The first one we can remove since we're scrapping `evt.stopPropagation()`. The other will just have to be modified a little to accommodate the new (less greedy) event logic. If you could also fix the spelling mistake in the description of the second test it would be greatly appreciated!

`on click closese the dialog and notifies cancellation` should read: `on click _closes_ the dialog and notifies cancellation`

Thanks again

**Comment by jonathanzong** [comment]
Upon further inspection I'm realizing that the intent of the removed test is to make sure click events don't propagate to items behind / unrelated to the dialog (see your comment [here](https://github.com/material-components/material-components-web/issues/225#issuecomment-296836508) in #225). My solution is to stop propagation when the `evt.currentTarget` (as opposed to `evt.target`) reaches the dialog root. I think this is the approach that lets events bubble, but not too far.

However, I'm not sure how to write a test for this behavior (how to mock events both inside and outside the dialog root). Do you have any advice or resources that might point me in the right direction?

**Comment by jonathanzong** [comment]
Hi @amsheehan, I just updated the tests and the CI checks have passed, would appreciate your feedback!

**Comment by Garbee** [comment]
> My solution is to stop propagation when the `evt.currentTarget` (as opposed to `evt.target`) reaches the dialog root. I think this is the approach that lets events bubble, but not too far.

This isn't emulating the natural dialog element though. Natural HTML `<dialog>`'s allow for the event to propagate all the way to the document itself. The component should reflect that functionality to align with how the web works.

File: packages/mdc-dialog/foundation.js [hunk 1]
```diff
@@ -52,7 +52,14 @@ export default class MDCDialogFoundation extends MDCFoundation {
   constructor(adapter) {
     super(Object.assign(MDCDialogFoundation.defaultAdapter, adapter));
     this.isOpen_ = false;
-    this.componentClickHandler_ = () => this.cancel(true);
+    this.componentClickHandler_ = (evt) => {
+      if (this.adapter_.eventTargetHasClass(evt.currentTarget, cssClasses.ROOT)) {
```
Comment (amsheehan) [inline packages/mdc-dialog/foundation.js]: As Garbee mentioned, events should be allowed to propagate down to the document.

File: test/unit/mdc-dialog/foundation.test.js [hunk 2]
```diff
@@ -185,14 +185,17 @@ test('#cancel calls notifyCancel when shouldNotify is set to true', () => {
   td.verify(mockAdapter.notifyCancel());
 });
 
-test('on dialog surface click calls evt.stopPropagation() to prevent click from propagating to background el', () => {
+test('on dialog surface click calls evt.stopPropagation() to prevent click from propagating past the root', () => {
```
Comment (amsheehan) [inline test/unit/mdc-dialog/foundation.test.js]: Once the event is allowed to reach the document, this test should be removed.

File: test/unit/mdc-dialog/foundation.test.js [hunk 3]
```diff
@@ -246,17 +249,40 @@ test('on dialog surface click does not close or notify if the event target is no
...
 
   foundation.open();
-  handlers.click();
+  handlers.click(evt);
 
   td.verify(mockAdapter.removeClass(cssClasses.OPEN));
   td.verify(mockAdapter.notifyCancel());
 });
 
+test('on click does not close or notify cancellation if event target is not the backdrop', () => {
```
Comment (amsheehan) [inline test/unit/mdc-dialog/foundation.test.js]: nit: might be worth mentioning for future devs that this effectively means a click on the surface does not close the dialog

File: test/unit/mdc-dialog/foundation.test.js [hunk 4]
```diff
@@ -246,17 +249,40 @@ test('on dialog surface click does not close or notify if the event target is no
...
+    target: {},
+  };
+
+  td.when(mockAdapter.eventTargetHasClass(evt.target, cssClasses.BACKDROP)).thenReturn(false);
+
+  foundation.open();
+  handlers.click(evt);
+
+  td.verify(mockAdapter.removeClass(cssClasses.OPEN), {times: 0});
+  td.verify(mockAdapter.notifyCancel(), {times: 0});
```
Comment (amsheehan) [inline test/unit/mdc-dialog/foundation.test.js]: It will be enough to test that the `removeClass()` adapter method isn't called.

**Comment by jonathanzong** [comment]
Thanks @amsheehan, yes I did sanity check this on Chrome/Safari/Firefox and events do propagate.

**Pull request merged**